Feature/ethernet or wifi support

kretzp opened the pull request:
Features:
- based on the work of mobremski
- Ethernet or Eifi for ESP32, (xor not both at same time) compile with -DHAS_ETHERNET
- Configurable via Webgui or Command WIFI OR ETHERNET on Hardware Page
- Note: Change Switch between WIFI and ETHERNET requires reboot to activate
- ESPEASY P2P works with both Ethernet or Wifi
- Note: Be aware with ESPEasyP2P Network, since IP Address will change. There could be conflicts.
- Mqtt works with both Ethernet or Wifi
- Ethernet shown in JSON in Sys:
"Ethernet":{
"Eth Wifi Mode":"ETHERNET",
"Eth connected":1,
"Eth Mode":"Full Duplex",
"Eth Speed":"100Mbps",
"Eth State":"Link Up",
"Eth Speed State":"Link Up Full Duplex 100Mbps"
- Ethernet usable ins System variables:
Ethernet
%ethwifimode% | ETHERNET | ETHERNET
%ethconnected% | 1 | 1
%ethduplex% | Full Duplex | Full%20Duplex
%ethspeed% | 100Mbps | 100Mbps
%ethstate% | Link Up | Link%20Up
%ethspeedstate% | Link Up Full Duplex 100Mbps | Link%20Up%20Full%20Duplex%20100Mbps
- Ethernet shown in System Info Page
Ethernet
Eth State: | Link Up
Eth Speed: | 100Mbps
Eth Mode: | Full Duplex
Eth MAC: | 30:AE:A4:9F:92:B8
Eth IP / Subnet: | 192.168.39.151 / 255.255.255.0
Eth Gateway: | 192.168.39.1
Eth DNS: | 192.168.39.1
- Ethernet Configuration and switch between Ethernet and Wifi is possible via commands
- Webserver JSON also updated with Ethernet Information
- New ESPEasyNetwork.cpp with Methods wrapping around ESPEasyWifi.cpp and ESPEasyEth.cpp, no ino but cpp/h files
- EPSEasyWifi*.ino converted to ESPEasyWifi*.cpp/h

kretzp committed:
atlas Sensor for pH an Redox (ORP) added

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

mobrembski committed:
Initial Ethernet support

mobrembski committed:
Added displaying Ethernet parameter in sysinfo and root page

mobrembski committed:
Added configuration of Eth Phy via Hardware settings web page

mobrembski committed:
Added Olimex ESP32-PoE board to DeviceModel and default settings

mobrembski committed:
Increased BUILD number to 20107

mobrembski committed:
Fix build on devices without Ethernet.

Forget to adjust size of GpioFactorySettingsStruct.

mobrembski committed:
Remove ifdefs for Ethernet in settings

mobrembski committed:
Renamed functions in ESPEasyEth, added eth* prefix

mobrembski committed:
Moved WifiGetHostname into more generic createRFCCompliantHostname

mobrembski committed:
Added basic implementation of Ethernet Static IP

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

kretzp committed:
- Atlas EZO new webserver added
- Serial --> logging
- No [TESTING]

kretzp committed:
Merge branch 'feature/ethernet_support' of https://github.com/mobrembski/ESPEasy into mega

kretzp committed:
Extended the great work from Michał Obrembski mobrembski, so that:
- mqtt works
- WifiConnected works as EthernetConnected

kretzp committed:
- Moved ETHEvent from ESPEasy.ino to ESPEasyEthEvent.ino
- Corrected Hostname Setting
- ESPEasy P2P works now with some minor changes for IP-Address Handling in Network.ino

kretzp committed:
Change requested by TD-er review: clarification of build file
https://github.com/letscontrolit/ESPEasy/pull/2962/files/8ba2bb88eadb98c4691cbb1f40c4d5466cb7b5b8

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_support

kretzp committed:
ESP32 with Ethernet or Wifi, not both at the same time configurable via Hardware Page.
- ESP P2P is working
- Only ETH.beginPacket() with Standard Parameter is working at the moment, configuration will follow.
- Network.in not cpp/h files but this will be changed in future
- Some TODO: PKR: comments will be removed in the future, too

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

kretzp committed:
Removed some comments
Build now working with HAS_ETHERNET not defined

kretzp committed:
- Ethernet Commands
- remove useless comments
- ETH.config moved after ETH.begin
- Added note: Be aware with ESPEasyP2P Network, since IP Address will change. There could be conflicts.
- Ethernet JSON
- Ethernet Variables

kretzp committed:
- ETHEvents moved to WifiEvents, because they are the same type

kretzp committed:
ESPEasyEth.ino --> ESPEasyEth.cpp/h
Network.ino --> Network.cpp/h

kretzp committed:
ESPEasyWiFi_credentials.ino --> cpp/h
ESPEasyWiFi_ProcessEvent.ino --> cpp/h
ESPEasyWiFi.ino --> cpp/h

kretzp committed:
Network.h/cpp --> ESPEasyNetwork.h/cpp

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

kretzp committed:
removed upload flags

kretzp committed:
- New events: ETHERNET#Connected and ETHERNET#Disconnected

kretzp committed:
Fixed
https://travis-ci.org/github/letscontrolit/ESPEasy/builds/680698472
ci build error

kretzp committed:
Fixed ci compile bug
https://travis-ci.org/github/letscontrolit/ESPEasy/builds/680724129

kretzp committed:
Fixed bug reported by mobremski

kretzp committed:
- UDP / ESPEasyP2P now works for ETHERNET
- Hostname Handling refactored for better overview:
  - String NetworkGetHostNameFromSettings() returns Settings.getHostname()
  - String NetworkCreateRFCCompliantHostname() creates a RFCcompliant Hostname from Settings.cpp
  - String NetworkGetHostname() returns the Hostname prevoiusly set in ETH or Wifi dependong on Mode
- MacAddress Handling refactored
  - String NetworkMacAddress() returns the Mac addres of ETH or Wifi depending on Mode
  - String WifiSoftAPmacAddress() returns WiFi.softAPmacAddress(mac) as String

TD-er commented on src/ESPEasyWiFiEvent.cpp:
```diff
@@ -113,7 +113,7 @@ void WiFiEvent(system_event_id_t event, system_event_info_t info) {
     case SYSTEM_EVENT_ETH_START:
       addLog(LOG_LEVEL_INFO, F("ETH Started"));
       char hostname[40];
```
Don't create objects here in a case.
If you need to create objects when handling a case, then you _must_ add scope braces.

```c++
case bla:
{
  // do your stuff
  break;
}
```

kretzp replied:
OK, I'll change it.

kretzp committed:
Merge branch 'feature/ethernet_or_wifi_support' of https://github.com/kretzp/ESPEasy into feature/ethernet_or_wifi_support

TD-er commented on src/ESPEasyEth.cpp:
```diff
@@ -1,9 +1,10 @@
+#include "eth_phy/phy.h"
 
 bool ethUseStaticIP() {
   return Settings.ETH_IP[0] != 0 && Settings.ETH_IP[3] != 255;
```
This I don't get.
The last byte of an IP-address can be 255. It all depends on the subnet mask whether or not it is a broadcast address.

kretzp replied:
Yes that's right. I will change it.

kretzp replied:
But how to change it?
It is the same way than for wifi check of static IPs in ESPEasyWifi.cpp:
```
bool useStaticIP() {
  return Settings.IP[0] != 0 && Settings.IP[0] != 255;
}

```

Shoud I change both to
` return Settings.IP[0] != 0;`
?

TD-er replied:
You check the **first** byte against 0 and the **last** byte against 255.
The existing check tests the **first** byte.

kretzp replied:
Oh yes, sorry. I need som ne glasses.
I chaged it to
`bool ethUseStaticIP() {
  return Settings.ETH_IP[0] != 0 && Settings.ETH_IP[0] != 255;
}`

TD-er commented on src/ESPEasyEth.cpp:
```diff
@@ -98,6 +102,14 @@ void ethPrintSettings() {
+    uint8_t mac[6];
+    char macStr[18] = { 0 };
+    esp_eth_get_mac(mac);
+    sprintf(macStr, "%02X:%02X:%02X:%02X:%02X:%02X", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
```
There is already a `formatMAC` function in the code.

kretzp replied:
Oh yes you are right. I will change it.

kretzp replied:
This String function is not needed, because there is a 
`ETH.macAddress()`
which returns a String.
I changed it to the byte version of the function and now 
`uint8_t * NetworkMacAddressAsBytes(uint8_t* mac)`
returns either this in Ethernet mode or
`WiFi.macAddress(mac)`
in Wifi mode.

kretzp committed:
This String function is not needed, because there is a
ETH.macAddress()
which returns a String.
I changed it to the byte version of the function and now
uint8_t * NetworkMacAddressAsBytes(uint8_t* mac)
returns either this in Ethernet mode or
WiFi.macAddress(mac)
in Wifi mode.

kretzp committed:
Check changed as requested by TD-er

Commit:
addFormNote(F("Be aware with ESPEasyP2P Network, since IP Address will change. There could be conflicts.")); not used anymore because P2P UDP works good

Commit:
Merge branch 'feature/ethernet_or_wifi_support' of https://github.com/kretzp/ESPEasy into feature/ethernet_or_wifi_support

Commit:
removede personal upload flags

TD-er commented on platformio_esp32_envs.ini:
```diff
@@ -22,6 +22,8 @@ build_flags               = ${mqtt_flags.build_flags}
                             -DCONFIG_FREERTOS_ASSERT_DISABLE
                             -DCONFIG_LWIP_ESP_GRATUITOUS_ARP
                             -DCONFIG_LWIP_GARP_TMR_INTERVAL=30
+upload_port               = /dev/cu.wchusbserial14630
```
Don't set the upload port, especially not in the common section.
For example this port for sure does not exist on my Windows environment.

Also upload speed of 921600 is for sure not advised to be used on all boards.

kretzp replied:
Yes I know, this was an error of mine and I already changed it before you reviewed it.

kretzp committed:
_P124_Atlas_EZO_pH.ino and _P222_Atlas_EZO_ORP.ino deleted since ther are not in ESPEasy but in Playground. There is another Issu, to put them in ESPEasy mega

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support
Conflicts resolved

kretzp commented:
Does anyone already found the time for a test? In my installation with Olimex, it's working fine (MQTT and ESPEasyP2P included) since 5 days wirh several switches from WiFi to Ethernet and back.

TD-er commented:
Not yet as I have been occupied with other work the last few days.

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

- Adaption from PullRequest #3055 to Ethernet: needsWiFi --> needsNetwork

kretzp commented:
> Not yet as I have been occupied with other work the last few days.

Thanksfor information. In between I adapted th Ethernet branch to the pullrequest '#055 you pulled today

kretzp committed:
- Error in Network Name fpr WiFi fixed
- logging improvements

TD-er commented on src/src/DataStructs/ProtocolStruct.h:
```diff
@@ -28,7 +28,7 @@ struct ProtocolStruct
   bool     usesSampleSets : 1;
   bool     usesExtCreds   : 1;
-  bool     needsWiFi      : 1;
+  bool     needsNetwork   : 1;
```
Nice one :)

kretzp replied:
Thx :)

uzi18 commented:
maybe we should add info about almost all experimental options needs to reboot to acivate ;)

kretzp commented:
> maybe we should add info about almost all experimental options needs to reboot to acivate ;)

That's right. At the moment, I am working on change between WiFi and Ethernet without reboot.

bccrew commented:
I guess this code will also work for the Olimex ESP32-POE-ISO board?

kretzp commented:
> I guess this code will also work for the Olimex ESP32-POE-ISO board?

I think so, but Ididn't test ist with this board. I tested it with Olimex ESP32-EVB-EA-IND

TD-er committed:
Merge branch 'mega' into feature/ethernet_or_wifi_support

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

bccrew commented:
Peter, just for documentation I am posting the error I receive here with the Olimex ESP32-POE-ISO-IND board.

I tried changing PHY address from 0 to 1 but it still doesn't work.

INIT : Booting version:  (ESP32 SDK v3.2.3-14-gd3e562907)
29 : Info  : INIT : Free RAM:288884
29 : Info  : INIT : Cold Boot - Restart Reason: CPU0: Vbat power on reset CPU1:                                                                                                                                                              for APP CPU, reseted by PRO CPU
30 : Info  : FS   : Mounting...
56 : Info  : CRC  : No program memory checksum found. Check output of crc2.py
66 : Info  : CRC  : SecuritySettings CRC   ...OK
70 : Info  : INIT : ETH_WIFI_MODE:1 (ETHERNET)
142 : Info  : INIT : Free RAM:283740
142 : Info  : INIT : I2C
143 : Info  : INIT : SPI not enabled
345 : Info  : INFO : Plugins: 84 [Normal] [TEST ESP32] (ESP32 SDK v3.2.3-14-gd3e                                                                                                                                                             562907)
346 : Info  : Connect to:
346 : Info  : ETHERNET
347 E (1348) emac: Timed out waiting for PHY register 0x2 to have value 0x0007(m                                                                                                                                                             ask 0xffff). Current value 0xffff
E (2348) emac: Timed out waiting for PHY register 0x3 to have value 0xc0f0(mask                                                                                                                                                              0xfff0). Current value 0xffff
E (2348) emac: Initialise PHY device Timeout
: Info  : Eth Wifi mode: ETHERNET ETH: PHY Type: ETH_PHY_LAN8720 PHY Addr: 1 Eth                                                                                                                                                              Clock mode: ETH_CLOCK_GPIO0_IN MDC Pin: 23 MIO Pin: 18 Power Pin: 12
2354 : Info  : After ETH.begin
2362 : Info  : Webserver: start
2365 : Info  : OTA  : Arduino OTA enabled on port 3232
3699 : Info  : WD   : Uptime 0 ConnectFailures 0 FreeMem 213500 EthSpeedState No                                                                                                                                                              Ethernet

kretzp commented:
> Peter, just for documentation I am posting the error I receive here with the Olimex ESP32-POE-ISO-IND board.
> 
> I tried changing PHY address from 0 to 1 but it still doesn't work.
> 
> INIT : Booting version: (ESP32 SDK v3.2.3-14-gd3e562907)
> 29 : Info : INIT : Free RAM:288884
> 29 : Info : INIT : Cold Boot - Restart Reason: CPU0: Vbat power on reset CPU1: for APP CPU, reseted by PRO CPU
> 30 : Info : FS : Mounting...
> 56 : Info : CRC : No program memory checksum found. Check output of crc2.py
> 66 : Info : CRC : SecuritySettings CRC ...OK
> 70 : Info : INIT : ETH_WIFI_MODE:1 (ETHERNET)
> 142 : Info : INIT : Free RAM:283740
> 142 : Info : INIT : I2C
> 143 : Info : INIT : SPI not enabled
> 345 : Info : INFO : Plugins: 84 [Normal] [TEST ESP32] (ESP32 SDK v3.2.3-14-gd3e 562907)
> 346 : Info : Connect to:
> 346 : Info : ETHERNET
> 347 E (1348) emac: Timed out waiting for PHY register 0x2 to have value 0x0007(m ask 0xffff). Current value 0xffff
> E (2348) emac: Timed out waiting for PHY register 0x3 to have value 0xc0f0(mask 0xfff0). Current value 0xffff
> E (2348) emac: Initialise PHY device Timeout
> : Info : Eth Wifi mode: ETHERNET ETH: PHY Type: ETH_PHY_LAN8720 PHY Addr: 1 Eth Clock mode: ETH_CLOCK_GPIO0_IN MDC Pin: 23 MIO Pin: 18 Power Pin: 12
> 2354 : Info : After ETH.begin
> 2362 : Info : Webserver: start
> 2365 : Info : OTA : Arduino OTA enabled on port 3232
> 3699 : Info : WD : Uptime 0 ConnectFailures 0 FreeMem 213500 EthSpeedState No Ethernet

Perhaps this is because you are using a LAN8710, which will not work. I will see to get a board like yours.

kretzp committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ethernet_or_wifi_support

bccrew commented:
Hi @kretzp It works with changing the Ethernet Clock to "50MHz APLL Inverted Output on GPIO17" like you suggested.

LAN lights are working too on the board.



Btw, @TD-er @kretzp 

On the src/WebServer_ConfigPage.ino we are now talking about "IP Settings" and "Ethernet IP Settings". Is it bad to make a permanent change in the code to rename the first FormSubHeader to "WIFI IP Settings"?

![image](https://user-images.githubusercontent.com/8138958/84180831-7301e680-aa88-11ea-9f75-6db9b165ecca.png)

bccrew commented:
I forgot to mention that POE also works for my ESP32-POE-ISO-IND board with the following config:

![image](https://user-images.githubusercontent.com/8138958/84181726-a133f600-aa89-11ea-8066-f3ae4af9ea63.png)

TD-er commented:
> Is it bad to make a permanent change in the code to rename the first FormSubHeader to "WIFI IP Settings"?

Nope, I would say please do.
Until now we didn't have anything other then WiFi, so it was obvious what was meant here.
But there is no harm in clarifying it into `WiFi IP Settings`  (small letters `i`, as is used everywhere where "WiFi" is mentioned)

bccrew commented:
@TD-er @kretzp 

I'm rather noob when it comes to editing PR's but I think I managed to create a PR on Peter's fork :)

```
  addFormSubHeader(F("WiFi IP Settings"));

  addFormIPBox(F("ESP WiFi IP"),         F("espip"),      Settings.IP);
  addFormIPBox(F("ESP WiFi GW"),         F("espgateway"), Settings.Gateway);
  addFormIPBox(F("ESP WiFi Subnetmask"), F("espsubnet"),  Settings.Subnet);
  addFormIPBox(F("ESP WiFi DNS"),        F("espdns"),     Settings.DNS);
  addFormNote(F("Leave empty for DHCP"));

#ifdef HAS_ETHERNET
  addFormSubHeader(F("Ethernet IP Settings"));

  addFormIPBox(F("ESP Ethernet IP"),         F("espethip"),      Settings.ETH_IP);
  addFormIPBox(F("ESP Ethernet GW"),         F("espethgateway"), Settings.ETH_Gateway);
  addFormIPBox(F("ESP Ethernet Subnetmask"), F("espethsubnet"),  Settings.ETH_Subnet);
  addFormIPBox(F("ESP Ethernet DNS"),        F("espethdns"),     Settings.ETH_DNS);
  addFormNote(F("Leave empty for DHCP"));
```

This way it will look like this:

![image](https://user-images.githubusercontent.com/8138958/84186543-f9bac180-aa90-11ea-936c-ef45cf8712c7.png)

bccrew committed:
Change FormSubHeader to clarify difference between IP settings of LAN or WiFi

kretzp commented:
> I forgot to mention that POE also works for my ESP32-POE-ISO-IND board with the following config:
> 
> ![image](https://user-images.githubusercontent.com/8138958/84181726-a133f600-aa89-11ea-8066-f3ae4af9ea63.png)

That's fine

kretzp committed:
Merge pull request #1 from bccrew/patch-1

Change FormSubHeader to distinguish settings between IP of LAN or WiFi

TD-er commented:
Ah that's the true spirit of Open Source :)
A pull request inside a pull request, being merged before the main PR is merged.

kretzp commented:
Hmmm, now I have a continious integration error in env:test_beta_ESP8266_4M1M
The size of the program is to big.
Error: The program size (1045652 bytes) is greater than maximum allowed (1044464 bytes)
How to save 1188 Bytes?
The rest seems to compile.

TD-er commented:
I have seen almost all recent builds are hitting this.
So I have to re-arrange the build definitions. See: #2590

bccrew commented:
Still stable after 6 days running!

![image](https://user-images.githubusercontent.com/8138958/84742803-4f233100-afb1-11ea-937d-1377fc6648ea.png)

TD-er commented:
I will make a test build for this and would be nice if it could also be tested on ESP8266 nodes to see if anything gets broken.
I don't see anything strange in this PR and as far as I can see it should work fine along with the current implementation.
But I don't want to have it waiting to be merged very long as it may only complicate future merges if we wait longer.

bccrew commented:
Please send me a binary to test !

TD-er commented:
> Please send me a binary to test !

It is building :)

TD-er commented:
Test build for this PR is ready:
- [ESPEasy_ESP32_mega-20200608-52-PR_3040.zip](https://www.dropbox.com/s/yk5zc9og0nj4hff/ESPEasy_ESP32_mega-20200608-52-PR_3040.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200608-52-PR_3040.zip](https://www.dropbox.com/s/cmmz3qnpex3vu76/ESPEasy_ESP82xx_mega-20200608-52-PR_3040.zip?dl=0)

kretzp commented:
> I will make a test build for this and would be nice if it could also be tested on ESP8266 nodes to see if anything gets broken.
> I don't see anything strange in this PR and as far as I can see it should work fine along with the current implementation.
> But I don't want to have it waiting to be merged very long as it may only complicate future merges if we wait longer.

I have it up and running on ESP8266 Nodes since the beginning and it runs fine. Everything works as expected.
![image](https://user-images.githubusercontent.com/15065072/84870102-28333080-b07f-11ea-804a-4ee1decfaefa.png)

![image](https://user-images.githubusercontent.com/15065072/84870210-4b5de000-b07f-11ea-89fe-f6eb060a8fc8.png)

This runs on a sonoff 4ch r2

TD-er commented:
Can you also test the test build I made, since that one got merges from the mega branch into it, so it will be just like what would be built if this PR is merged in the mega branch.

kretzp commented:
> Can you also test the test build I made, since that one got merges from the mega branch into it, so it will be just like what would be built if this PR is merged in the mega branch.

Yes I will test your build. But I always merged the mega branch into mine, too. I will install your build.

TD-er commented:
Some playground plugins have been entered to this PR.
Can you take them out of this PR?
If they need to be entered in the mega branch, we should make a different PR for them (+ assign a new number to them)

kretzp commented:
> Some playground plugins have been entered to this PR.
> Can you take them out of this PR?
> If they need to be entered in the mega branch, we should make a different PR for them (+ assign a new number to them)

Yes I will do that. Which plugins do you mean? I have two EZO Plugins in my head but not in the branch Feature/ethernet or wifi support. I removed them weeks ago.

TD-er commented:
Ah I see.
I was looking through the open change requests and then noticed those files.
But that was probably a view on an old revision of this PR.

GitHub is not always too clear on whether you're looking at the latest revision/commit or not when browsing through the change requests.

kretzp commented:
> Ah I see.
> I was looking through the open change requests and then noticed those files.
> But that was probably a view on an old revision of this PR.
> 
> GitHub is not always too clear on whether you're looking at the latest revision/commit or not when browsing through the change requests.

OK fine. I took a look on it, and there are no additional plugin in this branch, so it should be OK.

kretzp committed:
Merge branch 'mega' into feature/ethernet_or_wifi_support

kretzp commented:
> Test build for this PR is ready:
> 
> * [ESPEasy_ESP32_mega-20200608-52-PR_3040.zip](https://www.dropbox.com/s/yk5zc9og0nj4hff/ESPEasy_ESP32_mega-20200608-52-PR_3040.zip?dl=0)
> * [ESPEasy_ESP82xx_mega-20200608-52-PR_3040.zip](https://www.dropbox.com/s/cmmz3qnpex3vu76/ESPEasy_ESP82xx_mega-20200608-52-PR_3040.zip?dl=0)

Until now everything works fine on a Lolin Nodemcu v3 ESP8266

bccrew commented:
Here for me too on a D1 Mini !
